### PR TITLE
WIP - DO NOT MERGE - initial refactor xstate

### PR DESCRIPTION
**Pull request by swdpcomputing**

### CODE REVIEW

File: src/config/machines/transition-patterns.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ */
+export const createTransitions = (backTarget, nextTarget) => ({
+  on: {
+    BACK: {
+      target: backTarget,
+      actions: ['updateCurrentPageId']
+    },
+    NEXT: {
+      target: nextTarget,
+      actions: ['trackPageCompletion', 'updateCurrentPageId', 'updateAnswers']
```
Comment (DavidBarker-EE) [inline src/config/machines/transition-patterns.js]: This is assuming that all states require an answer update and page completion. Possibly, actions as defaulted arguments would be a good idea here.

File: src/config/machines/example-grant-machine.js [hunk 2]
```diff
@@ -99,30 +23,18 @@ export const exampleGrantMachine = createMachine({
...
-        NEXT: {
-          target: 'second-question', // NOSONAR:S1192 - need to improve this later
-          actions: ['trackPageCompletion', 'updateCurrentPageId', 'updateAnswers']
-        }
-      },
-
+      ...createTransitions('start', 'second-question'),
       meta: {
+        ...QUESTION_META,
         currentPageId: 'country',
```
Comment (DavidBarker-EE) [inline src/config/machines/example-grant-machine.js:35]: Like we have `createTransitions`, `createMeta` could be useful and using a lens we could potentially have something like `const [transitions, meta] = createQuestionState('country')('start', 'second-question')`.

Comment (swdpcomputing) [reply src/config/machines/example-grant-machine.js:35]: Agreed. This was on my deeper layer of refactor but as I said I got in a mess and decided to start afresh. Definitely one for the future.

File: src/config/machines/transition-patterns.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// src/config/machines/transition-patterns.js
+
+/**
+ * Creates standard transitions for a page in the journey
+ * @param {string} backTarget - The state to go back to
+ * @param {string} nextTarget - The state to go forward to
+ * @returns {object} Standard transition configuration
+ */
+export const createTransitions = (backTarget, nextTarget) => ({
```
Comment (Andrew-Folga) [inline src/config/machines/transition-patterns.js]: Great idea to extract it to separate file 👍 
Could we improve this to change the params to an object like so:


```
/**
 * Creates standard transitions for a page in the journey
 * @param {object} config - Object containing target states and actions
 * @param {string} config.back - The state to go back to
 * @param {string} config.next - The state to go forward to
 * @param {array} [config.backActions] - Actions for the BACK transition
 * @param {array} [config.nextActions] - Actions for the NEXT transition
 * @returns {object} Standard transition configuration
 */
export const createTransitions = ({
  back,
  next,
  backActions,
  nextActions]
}) => ({
  on: {
    BACK: {
      target: back,
      actions: backActions
    },
    NEXT: {
      target: next,
      actions: nextActions
    }
  }
});
```


and then usage:


```
country: {
  ...createTransitions({
    back: 'start',
    next: 'second-question',
    backActions: ['logBackNavigation', 'updateCurrentPageId'],
    nextActions: ['logNextNavigation', 'trackPageCompletion', 'updateCurrentPageId', 'updateAnswers']
  }),
  meta: {
    ...QUESTION_META,
    currentPageId: 'country',
    previousPageId: 'start',
    nextPageId: 'second-question',
    // other meta properties
  }
}
```

WDYT?

**Comment by Andrew-Folga** [comment]
LGTM but one improvement

File: src/config/machines/transition-patterns.js (hunk 3 above)
Comment (swdpcomputing) [reply src/config/machines/transition-patterns.js]: I've done a redo based on David's suggestions and yours. See if it works for you.

Comment (alanplatt) [reply src/config/machines/transition-patterns.js]: Does this not assume that all transitions are back and next? Maybe I'm missing something

File: src/config/machines/transition-patterns.js (hunk 1 above)
Comment (swdpcomputing) [reply src/config/machines/transition-patterns.js]: Edited.

**Comment by Andrew-Folga** [comment]
LGTM - will wait for tests and approve

File: src/config/machines/transition-patterns.js (hunk 3 above)
Comment (Andrew-Folga) [reply src/config/machines/transition-patterns.js]: Yeah I was thinking about that too but once we introduce new types of transitions we will improve?

Comment (Andrew-Folga) [reply src/config/machines/transition-patterns.js]: In fact I have some improvement in my coming PR to get rid of the DRY around `meta.nextPageId` and `on.NEXT.target` so I may as well address the different types of transition there too.

**Comment by Andrew-Folga** [comment]
Look like the XState Visulaiser plugin in VS Code (and I assume the online stately visualiser too) stops working with extracted chunks of logic - we need to look whether this can be fixed:

Screenshot 2024-11-08 at 13 39 57

**Review by Andrew-Folga — CHANGES_REQUESTED** [review]
Putting it on hold until the machine VS Code /stately Visualiser is fixed.

**Pull request closed without merging**